### ISSUE TITLE
PageTurns dont work in the FileBrowser anymore

### ISSUE THREAD

**Issue by hasezoey** [koreader/koreader#12908]
* KOReader version: 2024.11 / f025cce05925ca08b5e7134defdf2d0056f00ce2
* Device: Tolino Vision 5 (buttons)

#### Issue

PageTurns buttons dont work in the FileBrowser anymore, they worked in a previous version.
(they still work while reading a book though)

#### Steps to reproduce

0. Open koreader
1. Have the FileBrowser open (in any display mode) with enough files that multiple pages are available
2. Try to navigate the pages with the Page-Turn buttons
3. Observe nothing happening

##### `crash.log` (if applicable)
no crash, but logs when pressing the button in the FileBrowser:

(output from f025cce05925ca08b5e7134defdf2d0056f00ce2)
```txt
D/InputManager-JNI( 2487): [void android::NativeInputManager::handleInterceptActions(jint, nsecs_t, uint32_t&) 877]
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x648538f0,"Thread-230"] attached without supplying a name
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x64853b28,"Thread-231"] attached without supplying a name
D/PhoneWindowManager( 2487): mPstate.get() = 1
D/InputManager-JNI( 2487): [void android::NativeInputManager::handleInterceptActions(jint, nsecs_t, uint32_t&) 877]
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x64853e70,"Thread-232"] attached without supplying a name
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x648540a8,"Thread-233"] attached without supplying a name
D/PhoneWindowManager( 2487): mPstate.get() = 1
D/InputManager-JNI( 2487): [void android::NativeInputManager::handleInterceptActions(jint, nsecs_t, uint32_t&) 877]
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x648543f0,"Thread-234"] attached without supplying a name
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x64854628,"Thread-235"] attached without supplying a name
D/PhoneWindowManager( 2487): mPstate.get() = 1
D/InputManager-JNI( 2487): [void android::NativeInputManager::handleInterceptActions(jint, nsecs_t, uint32_t&) 877]
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x64854970,"Thread-236"] attached without supplying a name
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x64854ba8,"Thread-237"] attached without supplying a name
D/PhoneWindowManager( 2487): mPstate.get() = 1
D/InputManager-JNI( 2487): [void android::NativeInputManager::handleInterceptActions(jint, nsecs_t, uint32_t&) 877]
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x64854ef0,"Thread-238"] attached without supplying a name
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x64855128,"Thread-239"] attached without supplying a name
D/PhoneWindowManager( 2487): mPstate.get() = 1
D/InputManager-JNI( 2487): [void android::NativeInputManager::handleInterceptActions(jint, nsecs_t, uint32_t&) 877]
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x64855470,"Thread-240"] attached without supplying a name
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x648556a8,"Thread-241"] attached without supplying a name
D/PhoneWindowManager( 2487): mPstate.get() = 1
D/InputManager-JNI( 2487): [void android::NativeInputManager::handleInterceptActions(jint, nsecs_t, uint32_t&) 877]
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x648559f0,"Thread-242"] attached without supplying a name
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x64855c28,"Thread-243"] attached without supplying a name
D/PhoneWindowManager( 2487): mPstate.get() = 1
D/InputManager-JNI( 2487): [void android::NativeInputManager::handleInterceptActions(jint, nsecs_t, uint32_t&) 877]
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x64855f70,"Thread-244"] attached without supplying a name
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x648561a8,"Thread-245"] attached without supplying a name
D/PhoneWindowManager( 2487): mPstate.get() = 1
D/InputManager-JNI( 2487): [void android::NativeInputManager::handleInterceptActions(jint, nsecs_t, uint32_t&) 877]
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x648564f0,"Thread-246"] attached without supplying a name
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44271c70,peer=0x64856728,"Thread-247"] attached without supplying a name
D/PhoneWindowManager( 2487): mPstate.get() = 1
D/InputManager-JNI( 2487): [void android::NativeInputManager::handleInterceptActions(jint, nsecs_t, uint32_t&) 877]
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44207ad0,peer=0x64856a70,"Thread-248"] attached without supplying a name
W/art     ( 4187): Thread[10,tid=4201,Native,Thread*=0x44207ad0,peer=0x64856ca8,"Thread-249"] attached without supplying a name

```

---

Git bisection says fd3727274b44a7ba6531aa778189e0ab3a6e05fa is the first broken commit.
Coincidentally, it is also the commit that introduced #12904.

Bisecting `android-luajit-launcher` says [d1380ec3e5b780a1c91adcbdbce8492471c90009](https://github.com/koreader/android-luajit-launcher/commit/d1380ec3e5b780a1c91adcbdbce8492471c90009) is the first broken commit.
Which once again coincidentally, is also the commit that introduced #12904.

TL;DR: fd3727274b44a7ba6531aa778189e0ab3a6e05fa removed the handling for [`is_tolino`'s page-turn buttons](https://github.com/koreader/koreader/commit/fd3727274b44a7ba6531aa778189e0ab3a6e05fa#diff-f8e023b41718bacd7a310fcd56ca8215de5dfddbc2bb24c3a1ef33067f14993cL141-L146) and didnt seem to re-introduce them.

**Comment by hasezoey** [comment]
One workaround according to the wiki (if i understood the wiki and the commit message correctly), is:
```patch
diff --git a/frontend/device/android/event_map.lua b/frontend/device/android/event_map.lua
index 4cdf34a33..c9573dba6 100644
--- a/frontend/device/android/event_map.lua
+++ b/frontend/device/android/event_map.lua
@@ -9,8 +9,8 @@ return {
     [4] = "Back",   -- BACK
     [19] = "Up",    -- DPAD_UP
     [20] = "Down",  -- DPAD_UP
-    [21] = "Left",  -- DPAD_LEFT
-    [22] = "Right", -- DPAD_RIGHT
+    -- [21] = "Left",  -- DPAD_LEFT
+    -- [22] = "Right", -- DPAD_RIGHT
     [23] = "Press", -- DPAD_CENTER
     [24] = "LPgBack", -- VOLUME_UP
     [25] = "LPgFwd",  -- VOLUME_DOWN
@@ -31,4 +31,7 @@ return {
     [104] = "LPgBack", -- T68 PageUp
     [109] = "LPgFwd",  -- T68 PageDown
     [139] = "Menu",    -- T68 Menu
+
+    [21] = "LPgBack", -- Tolino page turn as dpad left/right
+    [22] = "LPgFwd",
 }
```

Alternatively i also tried using the following (and enabling `Use left and right keys for page turning`), but it didnt work:

```patch
diff --git a/frontend/device/android/device.lua b/frontend/device/android/device.lua
index 5f49892fa..20abda52d 100644
--- a/frontend/device/android/device.lua
+++ b/frontend/device/android/device.lua
@@ -73,7 +73,8 @@ local Device = Generic:extend{
     isAndroid = yes,
     model = android.prop.product,
     hasKeys = yes,
-    hasDPad = no,
+    hasDPad = yes,
+    useDPadAsActionKeys = yes,
     hasSeamlessWifiToggle = no, -- Requires losing focus to the sytem's network settings and user interaction
     hasExitOptions = no,
     hasEinkScreen = function() return android.isEink() end,
```
("it didnt work" as in even after a restart, it was still selecting entries instead of page-turning)

**Comment by pazos** [comment]
Removed non-aosp workarounds in frontend, as part of https://github.com/koreader/koreader/pull/12734

**Comment by Commodore64user** [comment]
> Also slightly less related, is the option Use left and right keys for page turning meant to do what i had intended it to do?
(i had tried it as the tolino vision 5's page turning keys are mapped to the dpad keycodes, at least guessing from the android event_map.lua)

No. That option, and the flags `hasDPad and useDPadAsActionKeys` are meant for [older] kindles, or people with actual keyboards. By default kindle [now] goes to previous/next chapter with left/right. That option changes it to previous/next page. 

I wouldn't recommend you to add them (`hasDPad and useDPadAsActionKeys`) as that will change a lot of behaviour in a lot of places.

**Comment by c-weider** [comment]
Same problem for me. But this helped me: https://gist.github.com/hasezoey/d16ba0f980f00cd2193132afe4714c7c#extra-koreader-patches
Thanks @hasezoey
